### PR TITLE
Improve JPA config trace and obfuscate passwords

### PR DESCRIPTION
**Pull request by dazey3**
for #15979

Signed-off-by: Will Dazey <dazeydev.3@gmail.com>

### CODE REVIEW

**Comment by dazey3** [comment]
#build

File: dev/com.ibm.ws.jpa.container.core/resources/com/ibm/ws/jpa/jpa.nlsprops
Comment (tkburroughs) [inline dev/com.ibm.ws.jpa.container.core/resources/com/ibm/ws/jpa/jpa.nlsprops]: "here" seems like it could be confusing; perhaps "is not recommended to be used in a EE server:" but perhaps wait for feedback from ID before changing.

Comment (dmuelle) [inline dev/com.ibm.ws.jpa.container.core/resources/com/ibm/ws/jpa/jpa.nlsprops]: - in general, we're advised to avoid making recommendations- can we instead just tell them not to use it?
- what does it mean that a "default" property is configured? Isn't the default what you get when you don't specify configuration? Also, the message calls it a "JPA default property" and the explanation says "JDBC property". If they're the same thing it's better to use one consistent term.

suggested edit:
 The following JDBC property, which is configured in the server.xml file, is not intended for Java or Jakarta EE applications: name = [{0}], value = [{1}].

File: dev/com.ibm.ws.jpa.container.core/resources/com/ibm/ws/jpa/jpa.nlsprops
Comment (dmuelle) [inline dev/com.ibm.ws.jpa.container.core/resources/com/ibm/ws/jpa/jpa.nlsprops]: IBM style guidance is to avoid modal "should"

suggested edit:
This JDBC property is not intended for Java or Jakarta EE applications and must be used only in Java SE environments.

File: dev/com.ibm.ws.jpa.container.core/resources/com/ibm/ws/jpa/jpa.nlsprops
Comment (dmuelle) [inline dev/com.ibm.ws.jpa.container.core/resources/com/ibm/ws/jpa/jpa.nlsprops]: Change the server or application configuration to use a JNDI data source and remove the property.

**Review by dmuelle — CHANGES_REQUESTED** [review]
ID review- let me know if you have any questions. Once the comments are addressed, I'll add the ID Reviewed label

File: dev/com.ibm.ws.jpa.container/src/com/ibm/ws/jpa/container/osgi/internal/JPAComponentImpl.java
Comment (tkburroughs) [inline dev/com.ibm.ws.jpa.container/src/com/ibm/ws/jpa/container/osgi/internal/JPAComponentImpl.java:224]: Did you consider leaving the passwords, or possibly all values as SerializableProtectedString, and only convert them when they are passed to the JPA provider, and/or only converting the non-passwords for trace?

Converting things later would be less efficient, but potentially safer.  I was also wondering, should we be protecting this list from modification by the JPA provider?  For example, what if the JPA provider added more properties to the list, or changed the list values, then all subsequent EMF creates would see the modified list, and not the original.  This could be a good reason to create a new list to pass on every create, and then convert from protected string at that time.

**Review by tkburroughs — COMMENTED** [review]
I think what you have is fine; pending ID approval; so my comments are just some thoughts to consider.

**Review by tkburroughs — APPROVED** [review]

File: dev/com.ibm.ws.jpa.container.core/resources/com/ibm/ws/jpa/jpa.nlsprops
Comment (dazey3) [reply dev/com.ibm.ws.jpa.container.core/resources/com/ibm/ws/jpa/jpa.nlsprops]: > what does it mean that a "default" property is configured?

Users configure "default" JPA properties in their server.xml configuration. These default properties become the defaults for all  container managed persistence units in all applications running on that server

Here is an example of a JPA JDBC property:
```xml
<server>
    <featureManager>
        ...
    </featureManager>

    <jpa>
        <defaultProperties>
            <property name="javax.persistence.jdbc.url" value="jdbc:mysql://localhost/testDB" />
        </defaultProperties>
    </jpa>
</server>
```
If a customer uses one of the `javax.persistence.jdbc.*` properties (referred to as JDBC properties), then we will log warning `CWWJP0057W` to inform them that this property should not be used in this way.

> we're advised to avoid making recommendations- can we instead just tell them not to use it?

Ill have to discuss with my team if we can ignore these properties instead of just warn against their use. If so, I can update the message accordingly

File: dev/com.ibm.ws.jpa.container.core/resources/com/ibm/ws/jpa/jpa.nlsprops
Comment (dazey3) [reply dev/com.ibm.ws.jpa.container.core/resources/com/ibm/ws/jpa/jpa.nlsprops]: How about:
```
    This JDBC property is being used in a Java EE environment, but is intended for use in Java SE environments only
```

File: dev/com.ibm.ws.jpa.container/src/com/ibm/ws/jpa/container/osgi/internal/JPAComponentImpl.java
Comment (dazey3) [reply dev/com.ibm.ws.jpa.container/src/com/ibm/ws/jpa/container/osgi/internal/JPAComponentImpl.java:224]: > Converting things later would be less efficient, but potentially safer.

That was the major reason. I didnt want to have to iterate all the properties again to convert before sending to the provider.

>I was also wondering, should we be protecting this list from modification by the JPA provider?

I think it is protected. If you read through `com.ibm.ws.jpa.management.JPAPUnitInfo.createEMFactory`, we create a temporary Map of properties, then add all the integration/default properties to that Map, finally sending it to the provider via `javax.persistence.spi.PersistenceProvider.createContainerEntityManagerFactory(PersistenceUnitInfo arg0, Map arg1)`. If the provider modifies that Map in their implementation, thats fine because we are done with it after passing it in

File: dev/com.ibm.ws.jpa.container.core/resources/com/ibm/ws/jpa/jpa.nlsprops
Comment (dmuelle) [reply dev/com.ibm.ws.jpa.container.core/resources/com/ibm/ws/jpa/jpa.nlsprops]: looks good- slight edit to move the "only"

    This JDBC property is being used in a Java EE environment, but is intended for use only in Java SE environments.

**Comment by dazey3** [comment]
@dmuelle I update the message incorporating your review comments. Let me know if there are any further refinements you believe would be beneficial. Thanks!

File: dev/com.ibm.ws.jpa.container.core/resources/com/ibm/ws/jpa/jpa.nlsprops
Comment (dmuelle) [inline dev/com.ibm.ws.jpa.container.core/resources/com/ibm/ws/jpa/jpa.nlsprops]: per IBM style, avoid future tense

The property will be ignored
--->
This property is ignored

Comment (dmuelle) [reply dev/com.ibm.ws.jpa.container.core/resources/com/ibm/ws/jpa/jpa.nlsprops]: slight edit suggestion

 A JPA default property that is configured in the server.xml file contains an empty name or empty value:

File: dev/com.ibm.ws.jpa.container.core/resources/com/ibm/ws/jpa/jpa.nlsprops
Comment (dmuelle) [inline dev/com.ibm.ws.jpa.container.core/resources/com/ibm/ws/jpa/jpa.nlsprops]: misplaced "only"

A JPA default property that is intended for use only in Java SE environments is configured in the server.xml file : name = [{0}], value = [{1}]. This property is ignored.

File: dev/com.ibm.ws.jpa.container.core/resources/com/ibm/ws/jpa/jpa.nlsprops
Comment (dmuelle) [inline dev/com.ibm.ws.jpa.container.core/resources/com/ibm/ws/jpa/jpa.nlsprops]: ...for use in Java SE environments only
--->
...for use only in Java SE environments

**Review by dmuelle — CHANGES_REQUESTED** [review]
looks good- a few more minor suggestions

**Comment by dazey3** [comment]
#build

**Review by dmuelle — APPROVED** [review]
LGTM thanks!

**Comment by dazey3** [comment]
#run-libby-bot

**Pull request merged**
